send-transaction-service/src/transaction_client.rs

```diff
@@ -0,0 +1,177 @@
+    leader_forward_count: u64,
+}
+
+// Manual implementation of Clone without requiring T to be Clone
+impl<T> Clone for ConnectionCacheClient<T>
+where
+    T: TpuInfo + std::marker::Send + 'static,
+{
+    fn clone(&self) -> Self {
+        Self {
+            connection_cache: Arc::clone(&self.connection_cache),
+            tpu_address: self.tpu_address,
+            tpu_peers: self.tpu_peers.clone(),
+            leader_info_provider: Arc::clone(&self.leader_info_provider),
+            leader_forward_count: self.leader_forward_count,
+        }
+    }
+}
+
+impl<T> ConnectionCacheClient<T>
+where
+    T: TpuInfo + std::marker::Send + 'static,
+{
+    pub fn new(
+        connection_cache: Arc<ConnectionCache>,
+        tpu_address: SocketAddr,
+        tpu_peers: Option<Vec<SocketAddr>>,
+        leader_info: Option<T>,
+        leader_forward_count: u64,
+    ) -> Self {
+        let leader_info_provider = Arc::new(Mutex::new(CurrentLeaderInfo::new(leader_info)));
+        Self {
+            connection_cache,
+            tpu_address,
+            tpu_peers,
+            leader_info_provider,
+            leader_forward_count,
+        }
+    }
+
+    fn get_tpu_addresses_with_slots<'a>(
+        &'a self,
+        leader_info: Option<&'a T>,
+    ) -> Vec<&'a SocketAddr> {
+        leader_info
+            .map(|leader_info| {
+                leader_info
+                    .get_leader_tpus(self.leader_forward_count, self.connection_cache.protocol())
+            })
+            .filter(|addresses| !addresses.is_empty())
+            .unwrap_or_else(|| vec![&self.tpu_address])
+    }
+
+    fn send_transactions(
+        &self,
+        peer: &SocketAddr,
+        wire_transactions: Vec<Vec<u8>>,
+        stats: &SendTransactionServiceStats,
+    ) {
+        let mut measure = Measure::start("send-us");
+        let conn = self.connection_cache.get_connection(peer);
+        let result = conn.send_data_batch_async(wire_transactions);
+
+        if let Err(err) = result {
+            warn!(
+                "Failed to send transaction transaction to {}: {:?}",
+                self.tpu_address, err
+            );
+            stats.send_failure_count.fetch_add(1, Ordering::Relaxed);
+        }
+
+        measure.stop();
+        stats.send_us.fetch_add(measure.as_us(), Ordering::Relaxed);
+        stats.send_attempt_count.fetch_add(1, Ordering::Relaxed);
+    }
+}
+
+impl<T> TransactionClient for ConnectionCacheClient<T>
+where
+    T: TpuInfo + std::marker::Send + 'static,
+{
+    fn send_transactions_in_batch(
+        &self,
+        wire_transactions: Vec<Vec<u8>>,
+        stats: &SendTransactionServiceStats,
+    ) {
+        // Processing the transactions in batch
+        let mut addresses = self
+            .tpu_peers
+            .as_ref()
+            .map(|addrs| addrs.iter().collect::<Vec<_>>())
+            .unwrap_or_default();
+        let mut leader_info_provider = self.leader_info_provider.lock().unwrap();
```

Comment:
probably, it doesn't worth it to hold the lock for so long but in the original version it was hold even longer, so here I just try not to make it worse.